src/main/java/com/google/devtools/build/lib/analysis/actions/CustomCommandLine.java

```diff
@@ -1277,53 +1227,68 @@ public CustomCommandLine evaluateTreeFileArtifacts(Iterable<TreeFileArtifact> tr
+   *                   {@link Builder#addPath(PathFragment)} entries. Output paths embedded in
+   *                   larger strings and added via {@link Builder#add(String)} or other variants
+   *                   must be handled separately.
+   */
   @Override
-  public ImmutableList<String> arguments(@Nullable ArtifactExpander artifactExpander)
-      throws CommandLineExpansionException, InterruptedException {
-    return argumentsInternal(artifactExpander);
-  }
-
-  private ImmutableList<String> argumentsInternal(@Nullable ArtifactExpander artifactExpander)
+  public ImmutableList<String> arguments(@Nullable ArtifactExpander artifactExpander,
+      PathMapper pathMapper)
       throws CommandLineExpansionException, InterruptedException {
     ImmutableList.Builder<String> builder = ImmutableList.builder();
     List<Object> arguments = rawArgsAsList();
     int count = arguments.size();
+    // Track the last scalar, non-path argument (e.g. "--javacopts") so that the PathMapper can
+    // heuristically map subsequent argument collections that contain paths.
+    String previousFlag = null;
     for (int i = 0; i < count; ) {
       Object arg = arguments.get(i++);
       if (arg instanceof TreeFileArtifactArgvFragment) {
         arg = substituteTreeFileArtifactArgvFragment((TreeFileArtifactArgvFragment) arg);
       }
       if (arg instanceof NestedSet) {
-        evalSimpleVectorArg(((NestedSet<?>) arg).toList(), builder);
+        evalSimpleVectorArg(((NestedSet<?>) arg).toList(), builder, pathMapper, previousFlag);
+        previousFlag = null;
       } else if (arg instanceof Iterable) {
-        evalSimpleVectorArg((Iterable<?>) arg, builder);
+        evalSimpleVectorArg((Iterable<?>) arg, builder, pathMapper, previousFlag);
+        previousFlag = null;
       } else if (arg instanceof ArgvFragment) {
         if (artifactExpander != null && arg instanceof TreeArtifactExpansionArgvFragment) {
           TreeArtifactExpansionArgvFragment expansionArg = (TreeArtifactExpansionArgvFragment) arg;
           expansionArg.eval(builder, artifactExpander);
         } else {
-          i = ((ArgvFragment) arg).eval(arguments, i, builder, getPathStripper());
+          i = ((ArgvFragment) arg).eval(arguments, i, builder, pathMapper);
         }
+        previousFlag = null;
       } else if (arg instanceof ActionInput) {
-        builder.add(getPathStripper().getMappedExecPathString((ActionInput) arg));
+        builder.add(pathMapper.getMappedExecPathString((ActionInput) arg));
+        previousFlag = null;
       } else if (arg instanceof PathFragment) {
-        builder.add(getPathStripper().map((PathFragment) arg).getPathString());
+        builder.add(pathMapper.map((PathFragment) arg).getPathString());
+        previousFlag = null;
       } else {
-        builder.add(CommandLineItem.expandToCommandLine(arg));
+        previousFlag = CommandLineItem.expandToCommandLine(arg);
+        builder.add(previousFlag);
```

Comment:
What is this additional logic to add previous flags here for? I notice new failures and the expanded command line looks funky. For example:

```
turbine_graal failed: error executing Turbine command (from target </some target>) tools/jdk/turbine_graal --output --output blaze-out/bin/java/com/google/apps/fava/request/libparam_key-hjar.jar --output_deps --output_deps ...
```

Notice how `--output` and `--output_deps` are added twice. Were you intended so that when processing `Iterables` argument to keep track whether or not it's `javacopts` so that `getMapFn` returns the correct `structuredArgsStripper`? This is totally speculation on my part based on the previous regression

Reply:
Yes, you guessed right :-) I pushed a new commit to simplify the logic here and move it closer to the explaining comment. 

The logic should not result in any additional flags being emitted. Both the analysis test I added as well as the integration test in https://github.com/bazelbuild/bazel/pull/18155 don't reproduce this flag duplication. Could this be something specific to Blaze? Do your aquery-based tests detect this issue?

Reply:
Our internal aquery test didn't detect this. I found this doing a random test that took a number of target within google as testing targets.

Reply:
Is this the same turbine action as in Bazel? How should we continue to debug this? If you could reproduce it in a test, that would be very helpful. If not, any kind of additional information on this failing action would be useful.

Reply:
ok I'll have a TODO to look into this. Since the new fix worked, I prioritized testing more to uncover new issues.